domain/application/state/resource.go

```diff
@@ -109,33 +157,141 @@ SELECT $resourceToAdd.uuid,
+		return errors.Capture(err)
+	}
+	if num != int64(len(resUUIDs)) {
+		return errors.Errorf("expected %d rows to be deleted, got %d", len(resUUIDs), num)
+	}
+
+	potentialResourceUUID, err := st.addPotentialFromResourceUUIDs(ctx, tx, charmSource, resUUIDs)
+	if err != nil {
+		return errors.Capture(err)
+	}
+	resUUIDs = append(resUUIDs, potentialResourceUUID...)
+
 	// Prepare SQL statement to link resource with application.
 	linkStmt, err := st.Prepare(`
 INSERT INTO application_resource (application_uuid, resource_uuid)
 VALUES ($linkResourceApplication.*)`, linkResourceApplication{})
 	if err != nil {
-		return errors.Capture(err)
+		return errors.Capture(errors.Errorf("linking resources: %w", err))
 	}
 
 	// Insert resources
-	appUUID := args.appID.String()
-	for _, res := range resources {
-		// Insert the resource.
-		if err = tx.Query(ctx, insertStmt, res).Run(); err != nil {
-			return errors.Errorf("inserting resource %q: %w", res.Name, err)
-		}
+	appUUID := appID.String()
+	for _, res := range resUUIDs {
 		// Link the resource to the application.
 		if err = tx.Query(ctx, linkStmt, linkResourceApplication{
-			ResourceUUID:    res.UUID,
+			ResourceUUID:    res,
 			ApplicationUUID: appUUID,
 		}).Run(); err != nil {
-			return errors.Errorf("linking resource %q to application %q: %w", res.Name, appUUID, err)
+			return errors.Errorf("linking resource %q to application %q: %w", res, appUUID, err)
 		}
 	}
 	return nil
 }
+
+// addPotentialFromResourceUUIDs creates potential resources for each
+// charmhub charm's resources from the resources created before the
+// application. Returns a list of resource uuids to be added to the
+// application resource link table.
+func (st *State) addPotentialFromResourceUUIDs(ctx context.Context,
+	tx *sqlair.TX,
+	charmSource charm.CharmSource,
+	resUUIDs uuids,
+) (uuids, error) {
+
+	// Only add potential resources for charmhub charms. See
+	// comment in buildResourcesToAdd for more info.
+	if charmSource != charm.CharmHubSource {
+		return nil, nil
+	}
+
+	potentialUUIDs := make(uuids, len(resUUIDs))
+
+	findStoreResourcesStmt, err := st.Prepare(`
+SELECT r.uuid AS &resourceToAdd.uuid,
+       r.charm_uuid AS &resourceToAdd.charm_uuid,
+       r.charm_resource_name AS &resourceToAdd.charm_resource_name,
+       rot.name AS &resourceToAdd.origin_type_name,
+       rs.name  AS &resourceToAdd.state_name
+FROM   resource AS r
+JOIN   resource_origin_type AS rot ON r.origin_type_id = rot.id
+JOIN   resource_state AS rs ON r.state_id = rs.id
+WHERE  r.uuid IN ($uuids[:])`, uuids{}, resourceToAdd{},
+	)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var potentialResources []resourceToAdd
+	err = tx.Query(ctx, findStoreResourcesStmt, resUUIDs).GetAll(&potentialResources)
+	if err != nil && !errors.Is(err, sql.ErrNoRows) {
+		return nil, errors.Errorf("finding store pending resources for application %w", err)
+	} else if errors.Is(err, sql.ErrNoRows) {
+		return nil, nil
+	}
```

Comment:
This is equivalent to:
```suggestion
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	} else if err != nil
		return nil, errors.Errorf("finding store pending resources for application %w", err)
	}
```